### PR TITLE
testing gnu container

### PR DESCRIPTION
**Pull request by mer-a-o**
To build the container:

```
cd spack-stack
source setup.sh
spack stack create ctr docker-ubuntu-gcc-openmpi --template=empty
cd envs/docker-ubuntu-gcc-openmpi
```

edit `spack.yaml` to add this to `specs: []`
```
 specs: [base-env@1.0.0, jedi-base-env@1.0.0 ~fftw, jedi-ewok-env@1.0.0, jedi-fv3-env@1.0.0,
    jedi-mpas-env@1.0.0, jedi-ufs-env@1.0.0, bacio@2.4.1,
    bison@3.8.2, bufr@11.7.0, ecbuild@3.6.5, eccodes@2.25.0, ecflow@5,
    eckit@1.19.0, ecmwf-atlas@0.29.0 ~trans ~fftw, eigen@3.4.0, esmf@8.3.0b09,
    fckit@0.9.5, flex@2.6.4, fms@release-jcsda, g2@3.4.5, g2tmpl@1.10.0,
    gftl-shared@1.5.0, hdf5@1.12.1, hdf@4.2.15, ip@3.3.3, jasper@2.0.32, jedi-cmake@1.3.0,
    libpng@1.6.37, mapl@2.12.3, nccmp@1.9.0.1, netcdf-c@4.8.1, netcdf-cxx4@4.3.1,
    netcdf-fortran@4.5.4, nlohmann-json-schema-validator@2.1.0, nlohmann-json@3.10.5,
    parallel-netcdf@1.12.2, parallelio@2.5.4, py-numpy@1.22.3, py-pandas@1.4.0, py-scipy@1.8.0,
    py-shapely@1.8.0, sp@2.3.3, udunits@2.2.28, w3nco@2.4.1,
    yafyaml@0.5.1, zlib@1.2.12, odc@1.4.5, crtm@v2.3-jedi.4]
```
Next:

```
spack containerize > Dockerfile
sudo docker build -t gnucontainer . 2>&1 | tee ./logdocker.txt
```

relate to: https://github.com/JCSDA-internal/AOP22/issues/167

### CODE REVIEW

**Comment by mer-a-o** [comment]
I was able to build a gnu container with spack. All the jedi-bundle tested passed except for 3 soca tests:
```
	1366 - test_soca_3dvarlowres_soca (Failed)
	1374 - letkf_split_observer_post (Failed)
	1375 - test_soca_letkf_split_solver (Failed)
```
failure in `letkf_split_observer_post` is related to not finding `ncks` command. 
The other two failures happen at the compare step.

**Comment by climbfuji** [comment]
> I was able to build a gnu container with spack. All the jedi-bundle tested passed except for 3 soca tests:
> 
> ```
> 	1366 - test_soca_3dvarlowres_soca (Failed)
> 	1374 - letkf_split_observer_post (Failed)
> 	1375 - test_soca_letkf_split_solver (Failed)
> ```
> 
> failure in `letkf_split_observer_post` is related to not finding `ncks` command. The other two failures happen at the compare step.

That's great! I think your container needs to install `nco` (we use version 5.0.6 in skylab-1.0.x).

**Comment by mer-a-o** [comment]
I'm testing mpas-bundle with the new container and I get an error when configuring mpas-model
```
-- Adding bundle project MPAS
-- Cloning MPAS from https://github.com/JCSDA-internal/MPAS-Model.git into /home/nonroot/mpas-bundle/MPAS...
-- /home/nonroot/mpas-bundle/MPAS retrieved.
-- Updating MPAS to head of BRANCH release-stable...
-- git fetch --all @ /home/nonroot/mpas-bundle/MPAS
-- git fetch --all --tags @ /home/nonroot/mpas-bundle/MPAS
-- git checkout release-stable @ /home/nonroot/mpas-bundle/MPAS
-- [OPTION] MPAS_CORES: init_atmosphere;atmosphere
-- [OPTION] MPAS_DOUBLE_PRECISION: ON
-- [OPTION] MPAS_PROFILE: OFF
-- [OPTION] MPAS_OPENMP: OFF
-- [OPTION] BUILD_SHARED_LIBS: ON
-- Found OpenMP_Fortran: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5") found components: Fortran 
-- Found MPI: TRUE (found version "3.1") found components: Fortran 
-- Found NetCDF: /opt/view/include (found version "4.8.1") found components: Fortran 
-- FindNetCDF [/home/nonroot/mpas-bundle/MPAS/cmake/Modules/FindNetCDF.cmake]:
--   - NetCDF_VERSION [4.8.1]
--   - NetCDF_PARALLEL [TRUE]
--   - NetCDF_C_CONFIG_EXECUTABLE [/opt/view/bin/nc-config]
--   - NetCDF::NetCDF_C [SHARED] [Root: /opt/view] Lib: /opt/view/lib/libnetcdf.so 
--   - NetCDF_Fortran_CONFIG_EXECUTABLE [/opt/view/bin/nf-config]
--   - NetCDF::NetCDF_Fortran [SHARED] [Root: /opt/view] Lib: /opt/view/lib/libnetcdff.so 
-- Found PnetCDF: /opt/view/bin/pnetcdf-config (found version "1.12.2") found components: Fortran 
-- FindPnetCDF:
--   - PnetCDF_VERSION [1.12.2]
--   - PnetCDF_PREFIX [/opt/software/linux-ubuntu20.04-haswell/gcc-9.4.0/parallel-netcdf-1.12.2-mkojzjqqbccur5n73fbi2h5qbwhjdobr]
--   - PnetCDF Components Found: Fortran;C;CXX
-- Found PIO: /opt/view  found components: Fortran C 
-- FindPIO:
--   - PIO_PREFIX [/opt/view]
--   - PIO Components Found: Fortran;C;SHARED
CMake Error at MPAS/CMakeLists.txt:67 (find_program):
  Could not find CPP_EXECUTABLE using the following names: cpp
```

**Comment by climbfuji** [comment]
> I'm testing mpas-bundle with the new container and I get an error when configuring mpas-model
> 
> ```
> -- Adding bundle project MPAS
> -- Cloning MPAS from https://github.com/JCSDA-internal/MPAS-Model.git into /home/nonroot/mpas-bundle/MPAS...
> -- /home/nonroot/mpas-bundle/MPAS retrieved.
> -- Updating MPAS to head of BRANCH release-stable...
> -- git fetch --all @ /home/nonroot/mpas-bundle/MPAS
> -- git fetch --all --tags @ /home/nonroot/mpas-bundle/MPAS
> -- git checkout release-stable @ /home/nonroot/mpas-bundle/MPAS
> -- [OPTION] MPAS_CORES: init_atmosphere;atmosphere
> -- [OPTION] MPAS_DOUBLE_PRECISION: ON
> -- [OPTION] MPAS_PROFILE: OFF
> -- [OPTION] MPAS_OPENMP: OFF
> -- [OPTION] BUILD_SHARED_LIBS: ON
> -- Found OpenMP_Fortran: -fopenmp (found version "4.5") 
> -- Found OpenMP: TRUE (found version "4.5") found components: Fortran 
> -- Found MPI: TRUE (found version "3.1") found components: Fortran 
> -- Found NetCDF: /opt/view/include (found version "4.8.1") found components: Fortran 
> -- FindNetCDF [/home/nonroot/mpas-bundle/MPAS/cmake/Modules/FindNetCDF.cmake]:
> --   - NetCDF_VERSION [4.8.1]
> --   - NetCDF_PARALLEL [TRUE]
> --   - NetCDF_C_CONFIG_EXECUTABLE [/opt/view/bin/nc-config]
> --   - NetCDF::NetCDF_C [SHARED] [Root: /opt/view] Lib: /opt/view/lib/libnetcdf.so 
> --   - NetCDF_Fortran_CONFIG_EXECUTABLE [/opt/view/bin/nf-config]
> --   - NetCDF::NetCDF_Fortran [SHARED] [Root: /opt/view] Lib: /opt/view/lib/libnetcdff.so 
> -- Found PnetCDF: /opt/view/bin/pnetcdf-config (found version "1.12.2") found components: Fortran 
> -- FindPnetCDF:
> --   - PnetCDF_VERSION [1.12.2]
> --   - PnetCDF_PREFIX [/opt/software/linux-ubuntu20.04-haswell/gcc-9.4.0/parallel-netcdf-1.12.2-mkojzjqqbccur5n73fbi2h5qbwhjdobr]
> --   - PnetCDF Components Found: Fortran;C;CXX
> -- Found PIO: /opt/view  found components: Fortran C 
> -- FindPIO:
> --   - PIO_PREFIX [/opt/view]
> --   - PIO Components Found: Fortran;C;SHARED
> CMake Error at MPAS/CMakeLists.txt:67 (find_program):
>   Could not find CPP_EXECUTABLE using the following names: cpp
> ```

Same as for gcc, gfortran etc I guess - probably needs an alias for cpp-10 ?

**Comment by mer-a-o** [comment]
@climbfuji adding cpp to /usr/bin and linking it to cpp-10 solved the issue. I'm getting a new error at `make` step:
```
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:28:22:

   28 |   INTEGER(ESMF_KIND_I8), INTENT(IN) :: numerator
      |                      1
Error: Symbol 'esmf_kind_i8' at (1) has no IMPLICIT type
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:29:22:

   29 |   INTEGER(ESMF_KIND_I8), INTENT(IN) :: denominator
      |                      1
Error: Symbol 'esmf_kind_i8' at (1) has no IMPLICIT type
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:58:22:

   58 |   INTEGER(ESMF_KIND_I8) :: numerator_i8, denominator_i8
      |                      1
Error: Symbol 'esmf_kind_i8' at (1) has no IMPLICIT type
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:26:42:

   26 | SUBROUTINE fraction_to_stringi8( numerator, denominator, frac_str )
      |                                          1
Error: Symbol 'numerator' at (1) has no IMPLICIT type
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:26:55:

   26 | SUBROUTINE fraction_to_stringi8( numerator, denominator, frac_str )
      |                                                       1
Error: Symbol 'denominator' at (1) has no IMPLICIT type
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:60:16:

   60 |   denominator_i8 = INT( denominator, ESMF_KIND_I8 )
      |                1
Error: Symbol 'denominator_i8' at (1) has no IMPLICIT type; did you mean 'denominator'?
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:59:45:

   59 |   numerator_i8 = INT( numerator, ESMF_KIND_I8 )
      |                                             1
Error: Symbol 'esmf_kind_i8' at (1) has no IMPLICIT type
/home/nonroot/mpas-bundle/MPAS/src/external/esmf_time_f90/MeatMod.F90:59:14:

   59 |   numerator_i8 = INT( numerator, ESMF_KIND_I8 )
      |              1
Error: Symbol 'numerator_i8' at (1) has no IMPLICIT type; did you mean 'numerator'?
make[2]: *** [MPAS/src/external/esmf_time_f90/CMakeFiles/esmf.dir/build.make:231: MPAS/src/external/esmf_time_f90/CMakeFiles/esmf.dir/MeatMod.F90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:18359: MPAS/src/external/esmf_time_f90/CMakeFiles/esmf.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by climbfuji** [comment]
I wonder if the version of mpas used in mpas-bundle requires a specific version of ESMF.

**Comment by mer-a-o** [comment]
I removed `esmf` files from `/opt/view` and rebuilt mpas-bundle. I'm getting a new error:
```
[ 18%] Linking CXX executable test_util_floatcompare
[ 18%] Building Fortran object ropp-ufo/src/CMakeFiles/ropp-ufo.dir/ropp_io/ropp/ropp_io_success.f90.o
[ 18%] Building Fortran object ropp-ufo/src/CMakeFiles/ropp-ufo.dir/ropp_io/thin/ropp_io_thin_sg.f90.o
[ 18%] Linking Fortran shared library ../../../lib/libmpas_framework.so
/usr/bin/ld: cannot find -lnetcdf
/usr/bin/ld: cannot find -lnetcdf
[ 18%] Built target test_util_floatcompare
collect2: error: ld returned 1 exit status
make[2]: *** [MPAS/src/framework/CMakeFiles/framework.dir/build.make:574: lib/libmpas_framework.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:18521: MPAS/src/framework/CMakeFiles/framework.dir/all] Error 2

/usr/bin/ld: cannot find -lnetcdf
/usr/bin/ld: cannot find -lnetcdf
[ 18%] Built target test_util_floatcompare
collect2: error: ld returned 1 exit status
make[2]: *** [MPAS/src/framework/CMakeFiles/framework.dir/build.make:574: lib/libmpas_framework.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:18521: MPAS/src/framework/CMakeFiles/framework.dir/all] Error 2
```

**Comment by climbfuji** [comment]
> I removed `esmf` files from `/opt/view` and rebuilt mpas-bundle. I'm getting a new error:
> 
> ```
> [ 18%] Linking CXX executable test_util_floatcompare
> [ 18%] Building Fortran object ropp-ufo/src/CMakeFiles/ropp-ufo.dir/ropp_io/ropp/ropp_io_success.f90.o
> [ 18%] Building Fortran object ropp-ufo/src/CMakeFiles/ropp-ufo.dir/ropp_io/thin/ropp_io_thin_sg.f90.o
> [ 18%] Linking Fortran shared library ../../../lib/libmpas_framework.so
> /usr/bin/ld: cannot find -lnetcdf
> /usr/bin/ld: cannot find -lnetcdf
> [ 18%] Built target test_util_floatcompare
> collect2: error: ld returned 1 exit status
> make[2]: *** [MPAS/src/framework/CMakeFiles/framework.dir/build.make:574: lib/libmpas_framework.so] Error 1
> make[1]: *** [CMakeFiles/Makefile2:18521: MPAS/src/framework/CMakeFiles/framework.dir/all] Error 2
> 
> /usr/bin/ld: cannot find -lnetcdf
> /usr/bin/ld: cannot find -lnetcdf
> [ 18%] Built target test_util_floatcompare
> collect2: error: ld returned 1 exit status
> make[2]: *** [MPAS/src/framework/CMakeFiles/framework.dir/build.make:574: lib/libmpas_framework.so] Error 1
> make[1]: *** [CMakeFiles/Makefile2:18521: MPAS/src/framework/CMakeFiles/framework.dir/all] Error 2
> ```

That could be an error in the mpas-bundle `CMakeLists.txt` not asking for the netcdf library (maybe because it was added implicitly through the ESMF dependency?)

**Comment by climbfuji** [comment]
How about we get this PR ready to merge, noting that the current version of the container builds jedi/fv3-bundle? Then we can make the necessary updates for mpas-bundle later.

**Comment by mer-a-o** [comment]
Sure, we can do that. Please note that mpas CI uses the same container as all the other CIs. So we cannot yet update the CI container.

**Comment by climbfuji** [comment]
> Sure, we can do that. Please note that mpas CI uses the same container as all the other CIs. So we cannot yet update the CI container.

Yes, that's totally fine. Since the generic container config in this PR doesn't contain the specs to install, we either need another container recipe that builds on the generic one, or have instructions somewhere (jedi-docs or here? tbd).

**Comment by mer-a-o** [comment]
@climbfuji This PR is ready.

File: configs/containers/docker-ubuntu-gcc-openmpi.yaml [hunk 1]
```diff
@@ -65,8 +65,9 @@ spack:
       spack:
         url: https://github.com/noaa-emc/spack
         ref: jcsda_emc_spack_stack
+        #ref: spack-stack-1.0.1
```
Comment (climbfuji) [inline configs/containers/docker-ubuntu-gcc-openmpi.yaml]: Please remove this line

**Review by climbfuji — CHANGES_REQUESTED** [review]
Please remove the unnecessary comment, and commit the change with `[skip ci]` in the commit message

**Review by climbfuji — APPROVED** [review]

**Review by srherbener — APPROVED** [review]
I'm not very expert at the container building process, but the code looks okay to me (plus it's been tested pretty thoroughly). Thanks for all the hard work on this!

**Pull request merged**